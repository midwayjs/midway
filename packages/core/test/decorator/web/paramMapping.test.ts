import {
  ALL,
  Body,
  createRequestParamDecorator,
  File,
  Files,
  getClassMetadata,
  Headers,
  INJECT_CUSTOM_PARAM,
  Param,
  Pipe,
  PipeTransform,
  Query,
  Session,
} from '../../../src';

@Pipe()
export class TestPipe implements PipeTransform {
  transform(value: any) {
    return value;
  }
}

@Pipe()
export class ErrorPipe implements PipeTransform {
  transform(value: any) {
    throw new Error('error');
  }
}

function Token() {
  return createRequestParamDecorator(ctx => {
    return ctx.request.headers.token;
  }, [TestPipe]);
}

function ErrorToken() {
  return createRequestParamDecorator(ctx => {
    return ctx.request.headers.token;
  }, {
    pipes: [ErrorPipe],
    throwError: true
  });
}

class Test {
  async doget(
    @Param('cc', ) cc: any,
    @Param('aa', [TestPipe]) aa: any,
    @Query('dd') dd: any,
    @Query('bb', [TestPipe]) query: any,
    @Body('body', [TestPipe]) body: any,
    @Headers('tt', [TestPipe]) tt: any,
    @File({ requireFile: true }, [TestPipe]) f: any,
    @Files() files: any,
    @Session(ALL, [TestPipe]) bb: any,
    @Token() token: any
  ) {}

  async dogetError(@ErrorToken() token: any) {}
}

describe('/test/web/paramMapping.test.ts', () => {
  it('paramMapping decorator should be ok', () => {
    const meta = getClassMetadata(INJECT_CUSTOM_PARAM, Test);
    expect(meta).toMatchSnapshot();
  });
});
