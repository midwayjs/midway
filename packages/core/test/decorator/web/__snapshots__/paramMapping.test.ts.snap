// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/test/web/paramMapping.test.ts paramMapping decorator should be ok 1`] = `
{
  "doget": [
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": [Function],
        "type": 12,
      },
      "options": {},
      "parameterIndex": 7,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": "common:all_value_key",
        "type": 4,
      },
      "options": {},
      "parameterIndex": 6,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": undefined,
        "type": 6,
      },
      "options": {},
      "parameterIndex": 5,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": {
          "requireFile": true,
        },
        "type": 5,
      },
      "options": {},
      "parameterIndex": 4,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": "tt",
        "type": 3,
      },
      "options": {},
      "parameterIndex": 3,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": "body",
        "type": 1,
      },
      "options": {},
      "parameterIndex": 2,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": "bb",
        "type": 0,
      },
      "options": {},
      "parameterIndex": 1,
      "propertyName": "doget",
    },
    {
      "impl": true,
      "key": "web:router_param",
      "metadata": {
        "pipes": undefined,
        "propertyData": "aa",
        "type": 2,
      },
      "options": {},
      "parameterIndex": 0,
      "propertyName": "doget",
    },
  ],
}
`;
